pkg/kn/commands/namespaced_test.go

```diff
@@ -143,71 +143,111 @@ func TestGetNamespaceFallback(t *testing.T) {
 
-	tempFile := filepath.Join(tempDir, "empty")
-	err = ioutil.WriteFile(tempFile, []byte(""), test.FileModeReadWrite)
-	assert.NilError(t, err)
+	t.Run("EmptyConfig", func(t *testing.T) {
+		// Invalid kubeconfig
+		tempFile := filepath.Join(tempDir, "empty")
+		err = ioutil.WriteFile(tempFile, []byte(""), test.FileModeReadWrite)
+		assert.NilError(t, err)
+
+		kp := &KnParams{KubeCfgPath: tempFile}
+		actual, err := kp.CurrentNamespace()
+		if isOpenShiftCI() {
+			// In-cluster config overrides the mocked one in OpenShift CI
+			assert.NilError(t, err)
+			assert.Equal(t, actual, os.Getenv("NAMESPACE"))
+		} else {
+			assert.Assert(t, err != nil)
+			assert.Assert(t, clientcmd.IsConfigurationInvalid(err))
+			assert.Assert(t, actual == "")
+		}
 
-	// Invalid kubeconfig
-	kp := &KnParams{KubeCfgPath: tempFile}
-	actual, err := kp.CurrentNamespace()
-	assert.Assert(t, err != nil)
-	assert.Assert(t, clientcmd.IsConfigurationInvalid(err))
-	assert.Assert(t, actual == "")
-
-	// Missing kubeconfig
-	kp = &KnParams{KubeCfgPath: filepath.Join(tempDir, "missing")}
-	actual, err = kp.CurrentNamespace()
-	assert.Assert(t, err != nil)
-	assert.ErrorContains(t, err, "can not be found")
-	assert.Assert(t, actual == "")
-
-	// Variable namespace override
-	kp = &KnParams{fixedCurrentNamespace: FakeNamespace}
-	actual, err = kp.CurrentNamespace()
-	assert.NilError(t, err)
-	assert.Equal(t, actual, FakeNamespace)
+	})
 
-	// Fallback to "default" namespace from mock kubeconfig
-	tempFile = filepath.Join(tempDir, "mock")
-	err = ioutil.WriteFile(tempFile, []byte(BASIC_KUBECONFIG), test.FileModeReadWrite)
-	assert.NilError(t, err)
-	kp = &KnParams{KubeCfgPath: tempFile}
-	actual, err = kp.CurrentNamespace()
-	assert.NilError(t, err)
-	assert.Equal(t, actual, "default")
+	t.Run("MissingConfig", func(t *testing.T) {
+		// Missing kubeconfig
+		kp := &KnParams{KubeCfgPath: filepath.Join(tempDir, "missing")}
+		actual, err := kp.CurrentNamespace()
+		assert.Assert(t, err != nil)
+		assert.ErrorContains(t, err, "can not be found")
+		assert.Assert(t, actual == "")
+	})
+
+	t.Run("MissingConfig", func(t *testing.T) {
+		// Variable namespace override
+		kp := &KnParams{fixedCurrentNamespace: FakeNamespace}
+		actual, err := kp.CurrentNamespace()
+		assert.NilError(t, err)
+		assert.Equal(t, actual, FakeNamespace)
+	})
+
+	t.Run("MockConfig", func(t *testing.T) {
+		// Fallback to "default" namespace from mock kubeconfig
+		tempFile := filepath.Join(tempDir, "mock")
+		err = ioutil.WriteFile(tempFile, []byte(BASIC_KUBECONFIG), test.FileModeReadWrite)
+		assert.NilError(t, err)
+		kp := &KnParams{KubeCfgPath: tempFile}
+		actual, err := kp.CurrentNamespace()
+		assert.NilError(t, err)
+		if isOpenShiftCI() {
+			// In-cluster config overrides the mocked one in OpenShift CI
+			assert.Equal(t, actual, os.Getenv("NAMESPACE"))
+		} else {
+			assert.Equal(t, actual, "default")
+		}
+	})
+}
+
+func isOpenShiftCI() bool {
+	return os.Getenv("OPENSHIFT_CI") == "true"
```

Comment:
I wonder whether if the issue wouldn't happen in any in-cluster run of the unit tests (not necessarily only on OpenShift), so that we check for the same file that the client lib looks up to decided whether it runs in-cluster ?

That would make the tests more robust in other environments, too, and we could avoid a reference to OpenShift here in Knative.

Reply:
Reworked to client-go `Possible()` style.